packages/@aws-cdk/aws-kinesisfirehose/lib/destination.ts

```diff
@@ -0,0 +1,99 @@
+import * as logs from '@aws-cdk/aws-logs';
+import { Construct } from 'constructs';
+import { IDeliveryStream } from './delivery-stream';
+import { CfnDeliveryStream } from './kinesisfirehose.generated';
+
+/**
+ * A Kinesis Data Firehose delivery stream destination configuration.
+ */
+export interface DestinationConfig {
+  /**
+   * Schema-less properties that will be injected directly into `CfnDeliveryStream`.
+   */
+  readonly properties: object;
+}
+
+/**
+ * Options when binding a destination to a delivery stream.
+ */
+export interface DestinationBindOptions {
+  /**
+   * The delivery stream.
+   */
+  readonly deliveryStream: IDeliveryStream;
+}
+
+/**
+ * A Kinesis Data Firehose delivery stream destination.
+ */
+export interface IDestination {
+  /**
+   * Binds this destination to the Kinesis Data Firehose delivery stream.
+   *
+   * Implementers should use this method to bind resources to the stack and initialize values using the provided stream.
+   */
+  bind(scope: Construct, options: DestinationBindOptions): DestinationConfig;
+}
+
+/**
+ * Generic properties for defining a delivery stream destination.
+ */
+export interface DestinationProps {
+  /**
+   * If true, log errors when data transformation or data delivery fails.
+   *
+   * If `logGroup` is provided, this will be implicitly set to `true`.
+   *
+   * @default true - errors are logged.
+   */
+  readonly logging?: boolean;
+
+  /**
+   * The CloudWatch log group where log streams will be created to hold error logs.
+   *
+   * @default - if `logging` is set to `true`, a log group will be created for you.
+   */
+  readonly logGroup?: logs.ILogGroup;
+}
+
+/**
+ * Abstract base class that destination types can extend to benefit from methods that create generic configuration.
+ */
+export abstract class DestinationBase implements IDestination {
+  private logGroups: { [logGroupId: string]: logs.ILogGroup } = {};
+
+  constructor(protected readonly props: DestinationProps = {}) { }
+
+  abstract bind(scope: Construct, options: DestinationBindOptions): DestinationConfig;
+
+  protected createLoggingOptions(
+    scope: Construct,
+    deliveryStream: IDeliveryStream,
+    streamId: string,
+  ): CfnDeliveryStream.CloudWatchLoggingOptionsProperty | undefined {
+    return this._createLoggingOptions(scope, deliveryStream, streamId, 'LogGroup', this.props.logging, this.props.logGroup);
+  }
+
+  private _createLoggingOptions(
```

Comment:
What's with the double level of indirection? Why not put the implementation directly in `createLoggingOptions()`?

Reply:
I agree this is weird because this PR doesn't have the other usage of the private function. This is this way because there will also be a "BackupLogGroup" for logging for the backup destination [See This](https://github.com/aws/aws-cdk/pull/15505/files#diff-de49be6d7608aede1539994b73893c571819827e911cc300a538e3664d7c3d63R335).

 But I can definitely remove the double indirection here, and then add it in the PR for the Backup destination. That will probably result in a cleaner implementation anyway.